File: python/p97/Interleaving97.py
class Solution(object):
    def isInterleave(self, s1, s2, s3):
        if len(s1) + len(s2) != len(s3) :
            return False

        self.rows = len(s1) + 1
        self.cols = len(s2) + 1
        self.table = [[False for i in range(self.cols)] for j in range(self.rows)]

        self.table[0][0] = True

        for m in range(1, self.rows):
            if s1[m-1] == s3[m-1]:
                self.table[m][0] = self.table[m-1][0]

        for n in range(1, self.cols):
            if s2[n-1] == s3[n-1]:
                self.table[0][n] = self.table[0][n-1]

        for m in range(1, self.rows):
            for n in range(1, self.cols):
                if s1[m-1] == s3[m+n-1]:
                    self.table[m][n] |= self.table[m-1][n]

                if s2[n-1] == s3[m+n-1]:
                    self.table[m][n] |= self.table[m][n-1]


        return self.table[ self.rows-1][ self.cols-1]

File: go/p97/interleaving97.go
package p97

func IsInterleave(s1 string, s2 string, s3 string) bool{
	return isInterleave(s1,s2,s3)
}

func isInterleave(s1 string, s2 string, s3 string) bool {
	if len(s1) + len(s2) != len(s3){
		return false
	}
	rows := len(s1) + 1
	cols := len(s2) + 1
	var res = make([][]bool, rows)
	for i :=0; i < rows; i++{
		res[i] = make( []bool, cols )
		for j := 0; j < cols; j ++ {
			res[i][j] = false
		}
	}

	res[0][0] = true

	for m := 1; m < rows; m ++{
		if s1[m-1] == s3[m-1] {
			res[m][0] = res[m-1][0]
		}
	}

	for n := 1; n < cols; n ++{
		if s2[n-1] == s3[n-1]{
			res[0][n] = res[0][n-1]
		}
	}

	for m := 1; m < rows; m ++ {
		for n := 1; n < cols ; n ++  {
			if s1[m-1] == s3[m+n-1]{
				res[m][n] = res[m-1][n]
				if res[m][n] {
					continue
				}
			}
			if s2[n-1] == s3[m+n-1]{
				res[m][n] = res[m][n-1]
				if res[m][n]{
					continue
				}

			}
		}
	}
	return res[rows-1][cols-1]
}
